src/Interpreters/DatabaseCatalog.cpp

```diff
@@ -327,20 +404,44 @@ DatabaseAndTable DatabaseCatalog::getTableImpl(
 
     DatabasePtr database;
     {
+        DatabaseNameHints hints(*this);
```

Comment:
This pattern, which is applied several times in the PR, where you generate hints before checking for error seems very bad for performance. Getting hints requires acquiring catalog mutexes and looping over arrays, so it should only happen when they are actually necessary (i.e. on error)

Reply:
I guess this is required due to mutex lock...

Reply:
Why is it required? What mutex?

Reply:
@tavplubix sorry, mistook it with something else...

Reply:
@tavplubix ...or I'm not? you can see that further lock is acquired before entering exception state...

Reply:
...at 409:
```
std::lock_guard lock{databases_mutex};
```

Reply:
But why would you do `DatabaseNameHints hints(*this);`, which will do a `database_catalog.getDatabases()` (acquiring mutex and iterating over all databases) if it's not necessary? If you have a lot of queries, lots of databases or lots of tables, you'll be wasting performance here for no reason.

Reply:
...maybe it makes sense to integrate `DatabaseNameHints` into `DatabaseCatalog` to avoid such complications?

Reply:
> Don't you agree that doing the work once is better than doing it twice?

what do you mean? `DatabaseNameHints` only capture `const DatabaseCatalog &`, then it calls `DatabaseCatalog::getDatabases()`...

Reply:
@tavplubix the creation of `DatabaseNameHints` is not a problem at all - it only captures `const DatabaseCatalog &` - the further call to `hints.getAllRegisteredNames()` is a problem - it calls to `DatabaseCatalog::getDatabases()` before exception happens. Introduction `Unlocked` function maybe a solution, but seems kinda ad-hoc... maybe makes sense to integrate `DatabaseNameHints` into `DatabaseCatalog` or make it friend...

Reply:
But you don't need DatabaseNameHints at all if there wasn't any error? Why build something that you don't need?

Look at a simpler example in getDatabase: 

https://github.com/ClickHouse/ClickHouse/pull/49801/files#diff-b5282f5a655be18aa9feb2674f0b271ccf9a892bb8e6fc1f98787d73a8bd2255R620-R627

In a fast path scenario (the database exists)

Before:
* Lock DatabaseCatalog::databases_mutex.
* Iterates over DatabaseCatalog::databases.
* Finds it, returns a pointer to the Database Class.
* Unlocks DatabaseCatalog::databases_mutex

Now:
* Creates a DatabaseNameHints (cheap).
* Calls DatabaseNameHints::getAllRegisteredNames.
  * Calls DatabaseCatalog::getDatabases()
     * Locks DatabaseCatalog::databases_mutex
     * Copies DatabaseCatalog::databases (a std::map)
     * Unlocks Locks DatabaseCatalog::databases_mutex
  * Takes the result of getDatabases() and iterates over it to remove temp databases (another copy)
* Calls DatabaseNameHints::getHints (cheap I guess)
* Lock DatabaseCatalog::databases_mutex.
* Iterates over DatabaseCatalog::databases.
* Finds it, returns a pointer to the Database Class.
* Unlocks DatabaseCatalog::databases_mutex


Notice how now:
a) There is an extra lock/unlock of the mutex
b) There is a copy DatabaseCatalog::databases which might be expensive if there are many databases.
c) Copies the database map again to filter temp tables (this could be avoidable but it's there).

And absolutely none of this steps are necessary for the common case (no error/exception).

Reply:
To be clear, since we wrote several messages at the same time. I don't think locking the mutex twice and generating the hints on error is a problem; doing it always (on success) is.

Reply:
@Algunenano well, locking a mutex twice is a problem called deadlock, I guess :) unless it's a recursive_mutex :)

Reply:
Yep. I meant lock-unlock, lock-unlock, not lock-lock-unlock-unlock.

Reply:
Hm, do we need the lock to create the exception when `databases.end() == it`? Seems like we can just unlock the mutex in this case and then create `DatabaseNameHints` and call `getAllRegisteredNames()`. Yes, it will not be atomic, but it's totally fine.

Reply:
> well, it's a lock_guard - you can't unlock it at will :)

@yakov-olkhovskiy, you can do smth like
```
{
    std::lock_guard lock{databases_mutex};
    auto it = databases.find(table_id.getDatabaseName());
    if (databases.end() != it)
        database = it->second;
}
if (!database)
{
    if (exception)
    {
        // create exception with hints
    }
    return {};
}
```
Or you can replace `lock_guard` with `unique_lock`, but in this case you will have to deal with TSA